pgvector pushdown with UNION having duplicate column names sometimes throws exceptions

Steps to reproduce:
```sql
-- pgvector
CREATE EXTENSION vector SCHEMA public;
CREATE SCHEMA tpch;

CREATE TABLE tpch.items (id bigserial PRIMARY KEY, embedding vector(3));
INSERT INTO tpch.items (embedding) VALUES ('[1.1, 2.2, 3.3]'), ('[4.4, 5.5, 6.6]'), ('[7.7, 8.8, 9.9]');

-- Trino
 SELECT id, embedding AS vector FROM postgresql.tpch.items
  UNION ALL
 (SELECT id, embedding AS vector FROM postgresql.tpch.items ORDER BY euclidean_distance(embedding, ARRAY[1.1, 2.2, 3.3]) LIMIT 1)
```

The above Trino statement's behavior isn't deterministic. It sometimes works, sometimes fails with the following exception. 
```sh
io.trino.spi.TrinoException: ERROR: column "embedding_2" does not exist
  Hint: Perhaps you meant to reference the column "items.embedding".
  Position: 58
	at io.trino.plugin.jdbc.JdbcRecordCursor.handleSqlException(JdbcRecordCursor.java:312)
	at io.trino.plugin.jdbc.JdbcRecordCursor.advanceNextPosition(JdbcRecordCursor.java:187)
	at io.trino.spi.connector.RecordPageSource.getNextPage(RecordPageSource.java:88)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:268)
	at io.trino.operator.Driver.processInternal(Driver.java:403)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:306)
	at io.trino.operator.Driver.tryWithLock(Driver.java:709)
	at io.trino.operator.Driver.process(Driver.java:298)
	at io.trino.operator.Driver.processForDuration(Driver.java:269)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:890)
	at io.trino.execution.executor.dedicated.SplitProcessor.run(SplitProcessor.java:77)
	at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.lambda$run$0(TaskEntry.java:201)
	at io.trino.$gen.Trino_testversion____20240828_115029_1.run(Unknown Source)
	at io.trino.execution.executor.dedicated.TaskEntry$VersionEmbedderBridge.run(TaskEntry.java:202)
	at io.trino.execution.executor.scheduler.FairScheduler.runTask(FairScheduler.java:172)
	at io.trino.execution.executor.scheduler.FairScheduler.lambda$submit$0(FairScheduler.java:159)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: java.sql.SQLException: ERROR: column "embedding_2" does not exist
  Hint: Perhaps you meant to reference the column "items.embedding".
  Position: 58
	at io.trino.plugin.jdbc.JdbcRecordCursor.advanceNextPosition(JdbcRecordCursor.java:167)
	... 21 more
Caused by: java.util.concurrent.ExecutionException: org.postgresql.util.PSQLException: ERROR: column "embedding_2" does not exist
  Hint: Perhaps you meant to reference the column "items.embedding".
  Position: 58
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at io.trino.plugin.jdbc.JdbcRecordCursor.advanceNextPosition(JdbcRecordCursor.java:163)
	... 21 more
Caused by: org.postgresql.util.PSQLException: ERROR: column "embedding_2" does not exist
  Hint: Perhaps you meant to reference the column "items.embedding".
  Position: 58
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2733)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2420)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:372)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:517)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:434)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:194)
	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:137)
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryStatement.wrapCall(OpenTelemetryStatement.java:304)
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryPreparedStatement.executeQuery(OpenTelemetryPreparedStatement.java:64)
	at io.trino.plugin.jdbc.JdbcRecordCursor.lambda$advanceNextPosition$1(JdbcRecordCursor.java:158)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	... 3 more
```

--- a/plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteVectorDistanceFunction.java
+++ b/plugin/trino-postgresql/src/main/java/io/trino/plugin/postgresql/rule/RewriteVectorDistanceFunction.java
@@ -124,12 +124,13 @@ public final class RewriteVectorDistanceFunction
         if (expression instanceof Call call && call.getFunctionName().equals(CAST_FUNCTION_NAME)) {
             ConnectorExpression argument = getOnlyElement(call.getArguments());
             if (argument instanceof Variable variable) {
-                JdbcTypeHandle typeHandle = ((JdbcColumnHandle) context.getAssignment(variable.getName())).getJdbcTypeHandle();
+                JdbcColumnHandle columnHandle = (JdbcColumnHandle) context.getAssignment(variable.getName());
+                JdbcTypeHandle typeHandle = columnHandle.getJdbcTypeHandle();
                 // TODO type.equals("vector") should be improved to support pushdown on vector type which is installed in other schemas
                 if (!typeHandle.jdbcTypeName().map(type -> type.equals("vector")).orElse(false)) {
                     return Optional.empty();
                 }
-                return Optional.of(new ParameterizedExpression(quoted(variable.getName()), ImmutableList.of()));
+                return Optional.of(new ParameterizedExpression(quoted(columnHandle.getColumnName()), ImmutableList.of()));
             }
             return Optional.empty();
         }

--- a/plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlVectorType.java
+++ b/plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlVectorType.java
@@ -20,6 +20,7 @@ import io.trino.testing.AbstractTestQueryFramework;
 import io.trino.testing.QueryRunner;
 import io.trino.testing.sql.TestTable;
 import io.trino.testing.sql.TestView;
+import org.junit.jupiter.api.RepeatedTest;
 import org.junit.jupiter.api.Test;
 
 import java.util.stream.IntStream;
@@ -376,6 +377,20 @@ final class TestPostgreSqlVectorType
                     .hasMessageContaining("infinite value not allowed in vector");
             assertThatThrownBy(() -> postgreSqlServer.execute("SELECT v <=> '[NULL]' FROM " + table.getName()))
                     .hasMessageContaining("invalid input");
+        }
+    }
+
+    @RepeatedTest(10) // Regression test for https://github.com/trinodb/trino/issues/23152
+    void testDuplicateColumnWithUnion()
+    {
+        try (TestTable table = new TestTable(postgreSqlServer::execute, "test_union", "(id int, v vector(3))")) {
+            postgreSqlServer.execute("INSERT INTO " + table.getName() + " VALUES (1, '[1,2,3]'), (2, '[4,5,6]')");
+
+            assertThat(query("" +
+                    "SELECT id FROM " + table.getName() +
+                    " UNION ALL " +
+                    "(SELECT id FROM " + table.getName() + " ORDER BY cosine_distance(v, ARRAY[4,5,6]) LIMIT 1)"))
+                    .matches("VALUES 1, 2, 2");
         }
     }
 }

